tests/shared/src/test/scala/cats/tests/HashMapSuite.scala

```diff
@@ -0,0 +1,253 @@
+      val leftKvs = left.iterator.toList
+      val rightKvs = right.iterator.toList
+      val distinctKvs = distinctBindings(leftKvs ++ rightKvs)
+      val both = left.concat(right)
+      assert(both === HashMap.fromSeq(leftKvs ++ rightKvs))
+      assert(both.size === distinctKvs.size)
+      distinctKvs.foreach { case (k, v) =>
+        assert(both.contains(k))
+        assert(both.get(k) === Some(v))
+        if (right.contains(k))
+          assert(both.get(k) === right.get(k))
+        else
+          assert(both.get(k) === left.get(k))
+      }
+    }
+  }
+
+  property("fromSeq consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromSeq(kvs)
+
+      kvs.foreach { case (k, _) =>
+        assert(hashMap.contains(k))
+      }
+    }
+  }
+
+  property("fromIterableOnce consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromIterableOnce(kvs.view)
+
+      kvs.foreach { case (k, _) =>
+        assert(hashMap.contains(k))
+      }
+    }
+  }
+
+  property("fromFoldable consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromFoldable(kvs)
+
+      kvs.foreach { case (k, _) =>
+        assert(hashMap.contains(k))
+      }
+    }
+  }
+
+  property("iterator consistent with contains") {
+    forAll { (hashMap: HashMap[Int, String]) =>
+      val iterated = mutable.ListBuffer[(Int, String)]()
+      hashMap.iterator.foreach { iterated += _ }
+      assert(iterated.toList.distinctBy(_._1) === iterated.toList)
+      assert(iterated.forall { case (k, _) => hashMap.contains(k) })
+    }
+  }
+
+  property("foreach consistent with contains") {
+    forAll { (hashMap: HashMap[Int, String]) =>
+      val foreached = mutable.ListBuffer[(Int, String)]()
+      hashMap.foreach { (i: Int, s: String) => foreached += ((i, s)) }
+      assert(foreached.toList.distinct === foreached.toList)
+      assert(foreached.forall { case (k, _) => hashMap.contains(k) })
+    }
+  }
+
+  property("foreach and iterator consistent") {
+    forAll { (hashMap: HashMap[Int, String]) =>
+      val iterated = mutable.ListBuffer[(Int, String)]()
+      val foreached = mutable.ListBuffer[(Int, String)]()
+      hashMap.iterator.foreach { iterated += _ }
+      hashMap.foreach { (i: Int, s: String) => foreached += ((i, s)) }
+      assert(foreached.forall(iterated.contains))
+      assert(iterated.forall(foreached.contains))
+    }
+  }
+
+  property("size consistent with iterator") {
+    forAll { (hashMap: HashMap[Int, String]) =>
+      var size = 0
+      hashMap.iterator.foreach { _ => size += 1 }
+      assert(hashMap.size === size)
+    }
+  }
+
+  property("size consistent with foreach") {
+    forAll { (hashMap: HashMap[Int, String]) =>
+      var size = 0
+      hashMap.foreach { case _ => size += 1 }
+      assert(hashMap.size === size)
+    }
+  }
```

Comment:
another test
```scala
forAll { (h1: HashMap[Int, String], h2: HashMap[Int, String]) =>
  if (h1.toString == h2.toString) assertEquals(h1, h2)
  if (h1.show === h2.show) assert(h1 === h2)
}
```
this could help us see if there is something weird going on where two maps print the same, but are actually different.

Reply:
In fact this would probably have caught that deletion bug even more easily